### PR TITLE
Add completion support for field access and xml attribute access exprs

### LINKED ISSUE #32262: Revisit (optional) field access completion for optional fields
**Description:**
$title.

**Describe your problem(s)**
Optional field access is not allowed as an lvalue. i.e., it cannot be used to update a field and field access should be used instead.

```ballerina
type Record record {|
    int optionalField?;
|};

public function main() {
    Record x = {};

    // Invalid.
    x?.optionalField = 1;

    // Valid.
    x.optionalField = 1;
}
```

But right now, when I use `.` and select the field, it always changes to optional field access.
```ballerina
type Record record {|
    int optionalField?;
|};

public function main() {
    Record x = {};

    x.o<cursor>
}
```

optional_field_access_bug

Accepting this adds
```ballerina
x?.optionalField
```

**Describe your solution(s)**
I suggest we go ahead with field access either

1. if it is on the LHS
```ballerina
x.o<cursor> = 1;
```

2. or when the context doesn't imply it is a read
```ballerina
x.o<cursor>
```

### LINKED ISSUE #32596: Member access expression does not provide any suggestion
**Description:**
Consider the bellow sample

Screenshot from 2021-09-07 12-38-53

No suggestions were provided for member access expression

```bal
public function main() {
    record {|string name; int[] marks; string id?;|} a = {
        name: "Dulmina",
        marks: [23, 99, 54]
    };
    io:println(a["id"].<cursor>);
}
```

### LINKED ISSUE #32609: Fix completion support for XML attribute access expression
**Description:**
$title
Spec: https://ballerina.io/spec/lang/master/#xml-attribute-access-expr

### PR DESCRIPTION
**Pull request by nadeeshaan**
## Purpose
Fixes #32262
Fixes #32609
Fixes #32596

## Approach

https://user-images.githubusercontent.com/1329674/131994144-533dbfcd-a554-4219-80b3-2184813e5832.mp4

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by nadeeshaan** [comment]
@kaviththiranga @NipunaMarcus please note this behaviour change.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/OptionalFieldAccessWriteAmbiguityResolver.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        this.context = context;
+    }
+
+    @Override
+    public Boolean transform(ExpressionStatementNode node) {
+        return Boolean.FALSE;
+    }
+
+    @Override
+    public Boolean transform(AssignmentStatementNode node) {
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/OptionalFieldAccessWriteAmbiguityResolver.java:54]: I think we should address `compound assignment statement` (https://ballerina.io/ballerina-spec/spec.html#section_7.14.3) as well. Ex: `rec.optionalField += <value>`.

File: language-server/modules/langserver-core/src/test/resources/completion/typedesc_context/config/table_typedesc5a.json [hunk 2]
```diff
@@ -4,14 +4,5 @@
...
-    {
-      "label": "recId",
-      "kind": "Field",
-      "detail": "int",
-      "sortText": "G",
-      "insertText": "recId",
-      "insertTextFormat": "Snippet"
-    }
-  ]
+  "items": []
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/resources/completion/typedesc_context/config/table_typedesc5a.json:7]: Not sure how this got affected. This isn't supposed to be changed. This is the source code for this scenario:
```
import ballerina/module1;

public const int TEST_CONST = 12;

public function main() {
    int[] testArr = [];
    table<TableRecord> key(uniqueKey, <cursor>)
    int testValue1 = 12;
}

type Key record {
    string uuid;
};

type TableRecord record {
    readonly int recId;
    string name;
    int age;
    readonly Key uniqueKey;
};
```
And cursor is at `<cursor>` and we should suggest all readonly required fields as a key.

**Review by IMS94 — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/test/resources/completion/typedesc_context/config/table_typedesc5a.json (hunk 2 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/test/resources/completion/typedesc_context/config/table_typedesc5a.json]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/FieldAccessContext.java [hunk 3]
```diff
@@ -54,24 +63,34 @@ public FieldAccessContext(Class<T> attachmentPoint) {
...
+            completionItems.addAll(this.getXmlAttributeAccessCompletions(ctx));
+        }
+        /*
+        expr being a field access expression and current node at cursor being qualified name reference is valid
+        for xml attribute access expression where the attribute name being a qualified name reference.
+        In this scenario we don't need to get the entries against the expr. This use case is handled above.  
+         */
+        if (!QNameReferenceUtil.onQualifiedNameIdentifier(ctx, ctx.getNodeAtCursor())) {
+            List<Symbol> symbolList = resolver.getVisibleEntries(expr);
+            //Add typegurad and foreach snippets.
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/FieldAccessContext.java:81]:
```suggestion
            //Add typeguard and foreach snippets.
```

**Review by IMS94 — APPROVED** [review]
LGTM! There's a suggestion to fix a typo.

**Review by IMS94 — DISMISSED** [review]

**Review by mohanvive — APPROVED** [review]

**Pull request merged**